ant-networking/src/relay_manager.rs

```diff
@@ -9,47 +9,105 @@
 use crate::driver::{BadNodes, NodeBehaviour};
 use itertools::Itertools;
 use libp2p::{
-    core::transport::ListenerId, multiaddr::Protocol, Multiaddr, PeerId, StreamProtocol, Swarm,
+    core::transport::ListenerId, multiaddr::Protocol, swarm::ConnectionId, Multiaddr, PeerId,
+    StreamProtocol, Swarm,
 };
 use rand::Rng;
-use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};
+use std::{
+    collections::{btree_map::Entry, BTreeMap, HashMap, HashSet, VecDeque},
+    time::SystemTime,
+};
 
 const MAX_CONCURRENT_RELAY_CONNECTIONS: usize = 4;
 const MAX_POTENTIAL_CANDIDATES: usize = 1000;
 
+/// We could get multiple incoming connections from the same peer through multiple relay servers, and only one of them
+/// would succeed. So we wait and collect all such connections peer 'from' peer, instead of just recording the
+///  success/failure for eachconnection.
+const MAX_DURATION_TO_TRACK_INCOMING_CONNECTIONS_PER_PEER: std::time::Duration =
+    std::time::Duration::from_secs(20);
+
+/// The connections from a single peer through multiple relay servers.
+type ConnectionsFromPeer = Vec<(PeerId, ConnectionId, SystemTime, Option<bool>)>;
```

Comment:
better have comments to explain which each field stands for or purpose to be used.
for example, what's the bool flag for and what if it is None ?